lib/block-supports/layout.php

```diff
@@ -262,23 +262,27 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
 /**
  * Renders the layout config to the block wrapper.
  *
- * @param  string $block_content Rendered block content.
- * @param  array  $block         Block object.
- * @return string                Filtered block content.
+ * @param  string   $block_content  Rendered block content.
+ * @param  array    $parsed_block   The parsed block.
+ * @param  WP_Block $block          Instance of WP_Block for the block being rendered.
+ * @return string                   Filtered block content.
  */
-function gutenberg_render_layout_support_flag( $block_content, $block ) {
-	$block_type     = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+function gutenberg_render_layout_support_flag( $block_content, $parsed_block, $block ) {
+	$block_type     = WP_Block_Type_Registry::get_instance()->get_registered( $block->name );
 	$support_layout = block_has_support( $block_type, array( '__experimentalLayout' ), false );
 
 	if ( ! $support_layout ) {
 		return $block_content;
 	}
 
+	// Use the WP_Block instance to access block attributes, since that class
+	// provides default attribute values, while the $parsed_block doesn't.
+	$attributes             = $block->attributes;
 	$block_gap              = gutenberg_get_global_settings( array( 'spacing', 'blockGap' ) );
 	$global_layout_settings = gutenberg_get_global_settings( array( 'layout' ) );
 	$has_block_gap_support  = isset( $block_gap ) ? null !== $block_gap : false;
 	$default_block_layout   = _wp_array_get( $block_type->supports, array( '__experimentalLayout', 'default' ), array() );
-	$used_layout            = isset( $block['attrs']['layout'] ) ? $block['attrs']['layout'] : $default_block_layout;
+	$used_layout            = isset( $attributes['layout'] ) ? $attributes['layout'] : $default_block_layout;
```

Comment:
I think we need to use `$parsed_block['attrs']['layout']` here instead of `$attributes['layout']`, because `$attributes['layout']` contains the default layout as well as the specific block's layout attributes.

Reply:
Ok I see the reason for the change now - new Group blocks no longer have the layout attribute explicitly set. The problem with doing that is that now we have no way of distinguishing a new Group with default constrained layout from a legacy Group with the previous default layout. 

We need this to work on the front end for all existing markup, so there has to be a way to tell these two types of Group apart. The only way I could find to do that was explicitly setting layout on all new Groups, which landed us in the present pickle 😅

I'm curious - could [this bit](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation/edit/index.js#L382) in the Nav block also have similar consequences? We might need to find a better way of setting attributes on new blocks.

Reply:
> The problem with doing that is that now we have no way of distinguishing a new Group with default constrained layout from a legacy Group with the previous default layout.

It does work in the editor though, so I'm hoping it's possible to replicate that in the PHP code.

> I'm curious - could [this bit](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation/edit/index.js#L382) in the Nav block also have similar consequences? We might need to find a better way of setting attributes on new blocks.

Possibly, I'm not sure of the absolute root cause yet, but it could do with some further investigation.

`useBlockSync` is the other bit of code that that results in the infinite looping. This bit here:
https://github.com/WordPress/gutenberg/blob/5cbfe819b098b5e89ae313f2c6fb7fd2ba430bf3/packages/block-editor/src/components/provider/use-block-sync.js#L141-L172

When `setControlledBlocks()` is called on line 162, it replaces all the inner blocks, and the group block's effect runs again. But the group block's effect then causes `useBlockSync` to run again. And so on. I think @youknowriad knows this code best so might have some insights.

Reply:
It only seems to work in the editor when first adding the block. If you save and refresh the page, the block loses the constrained layout 😬

Reply:
> The only way I could find to do that was explicitly setting layout on all new Groups, which landed us in the present pickle 😅

Just thinking about this a little more... if it winds up being too hard to infer the values from the PHP side, I suppose part of the challenge is that the concepts of a default value when inserted, and what constitutes a default value from serialized markup, are linked. I'm wondering if there could be a way (somehow) for us to explicitly set a default value when a block is inserted, that is different from what determines whether or not a value is serialized? Or to put it differently — to tease apart the two different concepts of `default` for attributes in `block.json`. That might require some API changes, though, so could be a bit of a rabbithole 😓🤔

Reply:
Do you mean setting default attributes explicitly on blocks, instead of having the "default" state be attribute-less?

Reply:
I'm going to try the alternative Riad suggested [here](https://github.com/WordPress/gutenberg/pull/42763#issuecomment-1246363659), if it works it should solve the problem 😅

Reply:
> Do you mean setting default attributes explicitly on blocks, instead of having the "default" state be attribute-less?

That's what I was idly wondering about, yeah... but 🤞 Riad's suggestion gives us a good way forward!